upstairs/src/lib.rs

```diff
@@ -2437,35 +2588,59 @@ impl CruciblePseudoFile {
  */
 impl Read for CruciblePseudoFile {
     fn read(&mut self, buf: &mut [u8]) -> IOResult<usize> {
-        let data = Buffer::from_slice(buf);
+        let _guard = self.rmw_lock.read().unwrap();
+
+        let mut span =
+            IOSpan::new(self.offset, buf.len() as u64, self.block_size);
 
-        let mut waiter = self.guest.read(self.offset, data.clone());
+        let mut waiter = span.read_affected_blocks_from_guest(&self.guest);
         waiter.block_wait();
 
+        span.read_from_blocks_into_buffer(buf);
+
         // TODO: for block devices, we can't increment offset past the
         // device size but we're supposed to be pretending to be a proper
         // file here
         self.offset += buf.len() as u64;
 
-        // TODO: is there a better way to do this fill?
-        {
-            let vec = data.as_vec();
-            for i in 0..buf.len() {
-                buf[i] = vec[i];
-            }
-        }
-
         Ok(buf.len())
     }
 }
 
 impl Write for CruciblePseudoFile {
     fn write(&mut self, buf: &[u8]) -> IOResult<usize> {
-        let mut data = BytesMut::with_capacity(buf.len());
-        data.put_slice(buf);
+        let mut span =
+            IOSpan::new(self.offset, buf.len() as u64, self.block_size);
 
-        let mut waiter = self.guest.write(self.offset, data.freeze());
-        waiter.block_wait();
+        /*
+         * Crucible's dependency system will properly resolve requests in the order they're sent
```

Comment:
I would word this as:

"Crucible's dependency system will properly resolve requests in the order they are received.."

We don't actually know who "sent" it first, all Crucible knows is when it received them.  It may
be exactly the same.